File: cpp/cpp_100-999/652.cc
// C++
// leetcode 652
// https://leetcode.cn/problems/find-duplicate-subtrees/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
	vector<TreeNode*> res;
	map<string, int> cnt;
	string dfs(TreeNode *root) {
		if (root == nullptr)
			return " ";
		string key = "";
		key += to_string(root -> val) + "_" + dfs(root -> left) + dfs(root -> right);
		cnt[key]++;
		if (cnt[key] == 2)
			res.push_back(root);
		return key;
	}

    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
    	dfs(root);
    	return res;
    }
};

File: Go/Go_100-999/652.go
// Go
// leetcode 652
// https://leetcode.cn/problems/find-duplicate-subtrees/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

import (
	"strconv"
)

func findDuplicateSubtrees(root *TreeNode) (res []*TreeNode) {
	var cnt = make(map[string]int)

	var dfs func(p *TreeNode) string
	dfs = func(p *TreeNode) string {
		// 返回 以结点 p 为根结点的树的结构为特征的一个值
		if p == nil {
			return " "
		}

		key := ""
		key += strconv.Itoa(p.Val) + "_" + dfs(p.Left) + dfs(p.Right)
		cnt[key]++
		if cnt[key] == 2 {
			res = append(res, p)
		}
		return key
	}

	dfs(root)
	return res
}